ADR/0033-upstreaming-integration-service.md

```diff
@@ -0,0 +1,270 @@
+containing only one element: the new pullspec.
+
+In the above example, the constructor will select only pipelineruns which are completed and which
+are associated with the app `my-application`. When triggered, it will extract the same image and
+source details that we extract today to supply to a new [TestSubject].
+
+### Release Initiation, as owned by the release-service
+
+For the purpose of this ADR, we will only say that release-service will take over the
+responsibility of initiating Releases. That will be described in more detail in a subsequent,
+forthcoming ADR.
+
+Note, in this pattern, [Release] construction looks a lot like [TestSubject] construction. The
+triggered system knows enough about the triggering system to optionally initiate itself.
+
+## Use Cases in Detail
+
+**Using integration-service as a remote API from Jenkins.** A hypothetical platform team might want
+to make use of [integration-service] from a jenkins pipeline. For example, perhaps their jenkins job
+performs CI on a git repository that contains pullspecs which should be mirrored to some environment
+of theirs. Before mirroring, they want to make sure that the same test works against dozens of
+different kubernetes configurations on different public clouds. That platform team first
+creates a single tekton pipeline that contains the logic to perform the test given a single cluster.
+They then configure dozens of [IntegrationTestScenario] resources, one for each kubernetes
+configuration they want to target. When a new set of pullspecs is proposed to their git repo,
+Jenkins responds. Jenkins then creates a single [TestSubject] in the kubernetes namespace,
+representing the new set of images under test. [integration-service] responds, inspecting all of the
+defined [IntegrationTestScenarios], it invokes the [deployment-target-operator] (from [Decoupling
+Deployment]) to provision each cluster configuration and it runs tests of the [TestSubject] against
+each. The aggregate results are stored in the `status` of the [TestSubject] for jenkins to monitor
+and report back to the platform team in CI.
+
+**Using integration-service with TestSubject automation.** A hypothetical platform team might want
+to make use of [integration-service] in conjunction with tekton pipelines that they use for
+building images. Imagine, every night they have a cron trigger which rebuilds their images using
+a static Dockerfile that installs the latest debian packages available to a base image and then
+rebuilds all of their utility images on that new base image. Their pipelines expose the pullspecs of
+all of those images as results. The platform team configures a [TestSubjectConstructor] which
+instructs the [test-subject-construction-controller] to watch all PipelineRuns in their namespace with
+a certain label. As those builds finish, the [test-subject-construction-controller] constructs new
```

Comment:
The TestSubjectConstructor will have to be created for every set of TestSubjects created. We are implying that there is a (potentially reused) dependency on some external pattern. If there are multiple types of artifacts that are produced across build pipelines, there are consistency requirements imposed on these pipelines by a decoupled integration service controller. 

Is there a way that we can improve this? Can we make there be a single explicit requirement/dependency that exists between the services (i.e. an Artifact CRD) instead of these implicit ones?

Reply:
> The TestSubjectConstructor will have to be created for every set of TestSubjects created

I worry we are miscommunicating here; I don't understand this.

The platform team creates the TestSubjectConstructor **once** as part of setting up their process. The constructor resource should stay in place monitoring for new PipelineRuns, guiding the creation of TestSubjects night after night.

Reply:
```
spec:
  selector:
    fields:
      match:
        kind: PipelineRun
        version: tekton.dev/v1
    labels:
      match:
        "appstudio.openshift.io/application": my-application
        "pipelinesascode.tekton.dev/state": completed
```
In this example, the TestSubjectConstructor is scoped to matching on a specific `appstudio.openshift.io/application`, so if multiple Applications are created (i.e. mapping to multiple TestSubjects) then there will need to be multiple TestSubjectConstructors.

One way to get around this would be to convert from the application being used as the selector to instead just use the application name in the extractor (i.e. to set the value of `LABELNAME`).

Even with this generalization, there are implicit requirements imposed upon the build PipelineRuns since the extractor is consistent across the board. This implicit requirement provides no feedback back to the user during the build because the decoupling of the integration service only consumes the PipelineRuns and does not return any feedback to it. The pattern creates an opportunity where a user might want to modify the build pipeline which results in a TestSubject not being properly created if the extracted fields are missing or change behavior.